### PR TITLE
[SuperTextField][mobile] Fix exception when updateTextAndSelection (Resolves #1229)

### LINKED ISSUE #1229: [SuperTextField] Calling updateTextAndSelection to clear the content of the field on iOS  throws an exception
When we try clear the content of a text field (e.g. by pressing the cross icon as shown in the screenshot below), we would call:
```dart
textController.updateTextAndSelection(
  text: AttributedText(),
  selection: const TextSelection.collapsed(offset: 0),
);
```

Screenshot 2023-07-11 at 10 50 56

On iOS this throws the following exception:
```
======== Exception caught by foundation library ====================================================
The following assertion was thrown while dispatching notifications for ImeAttributedTextEditingController:
'package:flutter/src/painting/text_painter.dart': Failed assertion: line 1281 pos 12: 'selection.isValid': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack: 
#2      TextPainter.getBoxesForSelection (package:flutter/src/painting/text_painter.dart:1281:12)
#3      RenderParagraph.getBoxesForSelection (package:flutter/src/rendering/paragraph.dart:955:25)
#4      RenderParagraphProseTextLayout.getCharacterBox (package:super_text_layout/src/text_layout.dart:321:34)
#5      _TextScrollViewState.getCharacterRectAtPosition (package:super_editor/src/infrastructure/super_textfield/infrastructure/text_scrollview.dart:272:21)
#6      TextScrollController.ensureExtentIsVisible (package:super_editor/src/infrastructure/super_textfield/infrastructure/text_scrollview.dart:995:20)
#7      SuperIOSTextFieldState._onTextOrSelectionChange (package:super_editor/src/infrastructure/super_textfield/ios/ios_textfield.dart:357:27)
#8      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:403:24)
#9      ImeAttributedTextEditingController._onInnerControllerChange (package:super_editor/src/infrastructure/super_textfield/input_method_engine/_ime_text_editing_controller.dart:197:5)
#10     ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:403:24)
#11     AttributedTextEditingController.updateTextAndSelection (package:super_editor/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart:164:5)
#12     ImeAttributedTextEditingController.updateTextAndSelection (package:super_editor/src/infrastructure/super_textfield/input_method_engine/_ime_text_editing_controller.dart:402:21)
#13     _defaultPopoverToolbarBuilder.<anonymous closure> (package:super_editor/src/infrastructure/super_textfield/ios/ios_textfield.dart:593:22)
#14     _InkResponseState.handleTap (package:flutter/src/material/ink_well.dart:1154:21)
#15     GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:275:24)
#16     TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:654:11)
#17     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:311:5)
#18     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:244:7)
#19     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:630:9)
#20     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:98:12)
#21     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:143:9)
#22     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#23     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:141:18)
#24     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:127:7)
#25     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:465:19)
#26     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:445:22)
#27     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:331:11)
#28     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:400:7)
#29     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:363:5)
#30     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:320:7)
#31     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:293:9)
#32     _invoke1 (dart:ui/hooks.dart:158:13)
#33     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:382:7)
#34     _dispatchPointerDataPacket (dart:ui/hooks.dart:91:31)
(elided 2 frames from class _AssertionError)
The ImeAttributedTextEditingController sending notification was: Instance of 'ImeAttributedTextEditingController'
====================================================================================================
```

This was tested on iOS and macOS and it is only reproducible on iOS.

### PR DESCRIPTION
**Pull request by angelosilvestre**
[SuperTextField][mobile] Fix exception when updateTextAndSelection. Resolves #1229

On mobile, clearing the text with the following code causes an assertion failure: 

```dart
textController.updateTextAndSelection(
  text: AttributedText(),
  selection: const TextSelection.collapsed(offset: 0),
);
```

The root cause is in `TextScrollView`. When the text or selection changes, we try to ensure that the selection extent is visible. To do that, in some places we call `getCharacterRectAtPosition` with a selection offset computed by `widget.textEditingController.text.text.length - 1`, which results in `-1` when the text is empty:

https://github.com/superlistapp/super_editor/blob/c6e2cfe5a0f0f0a097ab30707d6e337502cbc8b8/super_editor/lib/src/infrastructure/super_textfield/infrastructure/text_scrollview.dart#L990-L992

https://github.com/superlistapp/super_editor/blob/c6e2cfe5a0f0f0a097ab30707d6e337502cbc8b8/super_editor/lib/src/infrastructure/super_textfield/infrastructure/text_scrollview.dart#L1009-L1010

This causes an assertion failure in `TextPainter.getBoxesForSelection`.

This PR changes `ensureExtentIsVisible` and `ensureBaseIsVisible` to do nothing if the text is empty.

I was able to write a test that fails without this change.

### CODE REVIEW

**Comment by angelosilvestre** [comment]
@matthew-carroll Flutter lastest master broke a lot of goldens. Some colors seem to be changed.

File: super_editor/test/super_textfield/type_into_super_textfield_test.dart [hunk 1]
```diff
@@ -63,6 +65,42 @@ void main() {
 
         expect(find.text("a", findRichText: true), findsNothing);
       });
+
+      testWidgetsOnAllPlatforms('types after clearing text', (tester) async {
```
Comment (matthew-carroll) [inline super_editor/test/super_textfield/type_into_super_textfield_test.dart:69]: You solved a problem with auto-scrolling when text is empty. Shouldn't you write a test specifically to verify that auto-scrolling doesn't blow up when the text is empty?

Comment (angelosilvestre) [reply super_editor/test/super_textfield/type_into_super_textfield_test.dart]: Updated.

**Comment by angelosilvestre** [comment]
@matthew-carroll The analyzer is complaining because `textScaleFactor` seems to be deprecated on Flutter master:

```
  info • 'textScaleFactor' is deprecated and shouldn't be used. Use textScaler instead. Use of textScaleFactor was deprecated in preparation for the upcoming nonlinear text scaling support. This feature was deprecated after v3.12.0-2.0.pre • lib/src/text_layout.dart:200:50 • deprecated_member_use
```

Should we handle this in a separate ticket?

**Comment by matthew-carroll** [comment]
Sure. Please file a ticket for that. I'm not sure why that didn't trigger a Flutter test registry failure?

**Review by matthew-carroll — APPROVED** [review]
LGTM

**Comment by matthew-carroll** [comment]
@angelosilvestre Looks like we still have failures in `test_super_text_layout`

**Comment by angelosilvestre** [comment]
@matthew-carroll The failure is the analyzer complaining about the `textScaleFactor` usage.

**Comment by matthew-carroll** [comment]
Oh, ok

**Pull request merged**